pkg/validate/serialize.go

```diff
@@ -31,14 +58,128 @@ func Verify(yamlFileName string) error {
 // (1) the yaml file can not be converted to json.
 // (2) there is a problem while unmarshalling in go type.
 // Returns an object of type olm.ClusterServiceVersion.
-func unmarshal(rawYAML []byte) (*olm.ClusterServiceVersion, error) {
+func unmarshal(rawYAML []byte) (olm.ClusterServiceVersion, error) {
 
 	var csv olm.ClusterServiceVersion
 
-	if err := yaml.Unmarshal(rawYAML, &csv); err != nil {
-		return nil, fmt.Errorf("error parsing CSV list (JSON) : %s", err)
+	rawJson, err := yaml.YAMLToJSON(rawYAML)
+	if err != nil {
+		fmt.Printf("error parsing raw YAML to Json: %s", err)
+		return csv, err
+	}
+	if err := json.Unmarshal(rawJson, &csv); err != nil {
+		return csv, fmt.Errorf("error parsing CSV list (JSON) : %s", err)
+	}
+
+	return csv, nil
+}
+
+func csvInspect(val interface{}) []string {
+
+	fieldValue := reflect.ValueOf(val)
+
+	switch fieldValue.Kind() {
+	case reflect.Struct:
+		return checkMissingFields(fieldValue, "", []string{})
+	default:
+		return []string{"Error: input file is not a valid CSV."}
+	}
+}
+
+// Takes in a string slice and checks if a string (x) is present in the slice.
+// Return true if the string is present in the slice.
+func containsStrict(a []string, x string) bool {
+	for _, n := range a {
+		if x == n {
+			return true
+		}
 	}
+	return false
+}
+
+// Recursive function that traverses a nested struct passed in as reflect value, and reports for errors/warnings
+// in case of null struct field values.
+// Returns a log of errors as slice of strings.
+func checkMissingFields(v reflect.Value, parent string, log []string) []string {
+
+	for i := 0; i < v.NumField(); i++ {
+
+		fieldValue := v.Field(i)
 
-	return &csv, nil
+		tag := v.Type().Field(i).Tag.Get("json")
+		// Ignore fields that are subsets of a primitive field.
+		if tag == "" {
+			continue
+		}
 
+		fields := strings.Split(tag, ",")
+		isOptionalField := containsStrict(fields, "omitempty")
+
+		newParent := ""
+		if parent == "" {
+			newParent = v.Type().Field(i).Name
+		} else {
+			newParent = parent + "." + v.Type().Field(i).Name
+		}
+
+		switch fieldValue.Kind() {
+		case reflect.Struct:
+			emptyVal := isEmptyValue(fieldValue)
+			if emptyVal && isOptionalField {
+				log = append(log, fmt.Sprintf("WARNING: Optional Struct missing (%s).", newParent))
+				continue
+			} else if emptyVal && !isOptionalField {
+				if parent == "" {
+					log = append(log, fmt.Sprintf("ERROR: Mandatory Struct missing (%s).", newParent))
+				} else {
+					log = append(log, fmt.Sprintf("ERROR: Mandatory Struct missing (%s).", newParent))
+				}
+				continue
+			}
+			log = checkMissingFields(fieldValue, newParent, log)
+		default:
```

Comment:
most of this code is redundant. Maybe you should do the empty check first and then do the switch statement on the kind to see if you have to recursively call the function

Reply:
Done.